hack/generator/pkg/astmodel/storage_conversion_factories.go

```diff
@@ -125,6 +128,83 @@ func assignToOptionalType(
 	}
 }
 
+// assignFromOptionalType will handle the case where the source type may be missing (nil)
+//
+// <original> := <source>
+// if <original> != nil {
+//    <destination> = *<original>
+// } else {
+//    <destination> = <zero>
+// }
+func assignFromOptionalType(
+	sourceEndpoint *StorageConversionEndpoint,
+	destinationEndpoint *StorageConversionEndpoint,
+	conversionContext *StorageConversionContext) StorageTypeConversion {
+
+	// Require source to be optional
+	sourceOptional, sourceIsOptional := AsOptionalType(sourceEndpoint.Type())
+	if !sourceIsOptional {
+		return nil
+	}
+
+	// Require a conversion between the unwrapped type and our source
+	unwrappedEndpoint := sourceEndpoint.WithType(sourceOptional.element)
+	conversion, _ := createTypeConversion(unwrappedEndpoint, destinationEndpoint, conversionContext)
+	if conversion == nil {
+		return nil
+	}
+
+	local := sourceEndpoint.CreateLocal("", "Read")
+
+	return func(reader dst.Expr, writer func(dst.Expr) []dst.Stmt, generationContext *CodeGenerationContext) []dst.Stmt {
+
+		var cacheOriginal dst.Stmt
+		var actualReader dst.Expr
+
+		// If the value we're reading is a local or a field, it's cheap to read and we can skip
+		// using a local (which makes the generated code easier to read). In other cases, we want
+		// to cache the value in a local to avoid repeating any expensive conversion.
+
+		if _, ok := reader.(*dst.Ident); ok {
+			// reading a local variable
+			cacheOriginal = nil
+			actualReader = reader
+		} else if _, ok := reader.(*dst.SelectorExpr); ok {
+			// reading a field
+			cacheOriginal = nil
+			actualReader = reader
+		} else {
+			// Something else, so we cache the original
+			cacheOriginal = astbuilder.SimpleAssignment(
+				dst.NewIdent(local),
+				token.DEFINE,
+				reader)
+			actualReader = dst.NewIdent(local)
+		}
+
+		checkForNil := astbuilder.NotEqual(actualReader, dst.NewIdent("nil"))
+
+		// If we have a value, need to convert it to our destination type
+		writeActualValue := conversion(
+			astbuilder.Dereference(actualReader),
+			writer,
+			generationContext)
+
+		writeZeroValue := writer(
+			&dst.BasicLit{
+				Value: zeroValue(destinationEndpoint.Type(), conversionContext.types),
```

Comment:
Passing `types` to `zeroValue()` feels odd as it's only needed for the case where the type is a TypeName because we need to resolve it to find out whether it's an enum, alias, or something else.

It also feels odd to have `zeroValue()` here as a global func; I'm now wondering whether it has a natural place to live on `Types` - nicely scoped, easy to call, and has access to the set of available types if really needed.

Reply:
I changed my mind - have moved it onto `Type` and it seems to have come out cleanly.